Doc/library/sqlite3.rst

```diff
@@ -47,85 +47,179 @@ This document includes four main sections:
       PEP written by Marc-André Lemburg.
 
 
+.. We use the following practises for SQL code:
+   - UPPERCASE for keywords
+   - snake_case for schema
+   - single quotes for string literals
+   - singular for table names
+   - if needed, use double quotes for table and column names
+
 .. _sqlite3-tutorial:
 
 Tutorial
 --------
 
-To use the module, start by creating a :class:`Connection` object that
-represents the database.  Here the data will be stored in the
-:file:`example.db` file::
+In this tutorial, you will create a database of Monty Python movies
+using basic :mod:`!sqlite3` functionality.
+It assumes a fundamental understanding of database concepts,
+including `cursors`_ and `transactions`_.
+
+First, we need to create a new database and open
+a database connection to allow :mod:`!sqlite3` to work with it.
+Call :func:`sqlite3.connect` to to create a connection to
+the database :file:`tutorial.db` in the current working directory,
+implicitly creating it if it does not exist::
 
    import sqlite3
-   con = sqlite3.connect('example.db')
+   con = sqlite3.connect("tutorial.db")
 
-The special path name ``:memory:`` can be provided to create a temporary
-database in RAM.
+The returned :class:`Connection` object ``con``
+represents the connection to the on-disk database.
 
-Once a :class:`Connection` has been established, create a :class:`Cursor` object
-and call its :meth:`~Cursor.execute` method to perform SQL commands::
+In order to execute SQL statements and fetch results from SQL queries,
+we will need to use a database cursor.
+Call :meth:`con.cursor() <Connection.cursor>` to create the :class:`Cursor`::
 
    cur = con.cursor()
 
-   # Create table
-   cur.execute('''CREATE TABLE stocks
-                  (date text, trans text, symbol text, qty real, price real)''')
+Now that we've got a database connection and a cursor,
+we can create a database table ``movie`` with columns for title,
+release year, and review score.
+For simplicity, we can just use column names in the table declaration --
+thanks to the `flexible typing`_ feature of SQLite,
+specifying the data types is optional.
+Execute the ``CREATE TABLE`` statement
+by calling :meth:`cur.execute(...) <Cursor.execute>`::
+
+   cur.execute("CREATE TABLE movie(title, year, score)")
+
+.. Ideally, we'd use sqlite_schema instead of sqlite_master below,
+   but SQLite versions older than 3.33.0 do not recognise that variant.
+
+We can verify that the new table has been created by querying
+the ``sqlite_master`` table built-in to SQLite,
+which should now contain an entry for the ``movie`` table definition
+(see `The Schema Table`_ for details).
+Execute that query by calling :meth:`cur.execute(...) <Cursor.execute>`,
+assign the result to ``res``,
+and call :meth:`res.fetchone() <Cursor.fetchone>` to fetch the resulting row::
```

Comment:
I noticed that the lines are a bit inconsistent -- some are long, some are short.  Is this intentional or are you planning to reformat them to something like this before merging the PR?
```suggestion
We can verify that the new table has been created by querying the
``sqlite_master`` table built-in to SQLite, which should now contain
an entry for the ``movie`` table definition (see `The Schema Table`_
for details).  Execute that query by calling :meth:`cur.execute(...)
<Cursor.execute>`, assign the result to ``res``, and call
:meth:`res.fetchone() <Cursor.fetchone>` to fetch the resulting row::
```

Reply:
It is intentional. I try to use semantic line breaks wherever possible (unless I end up with very short lines). Especially starting each sentence on a new line can help reduce future diffs.

Reply:
Indeed, this is something I practice and encourage as well—ensuring the physical structure of the source text corresponds to its logical structure rather than being essentially arbitrary has major benefits not only for diffs/merges, but also is very beneficial when writing, editing and reviewing—its easier to spot sentences and clauses that are too long or too short and quickly grasp the structure of the text, there's a lot more flexibility to atomically modify one clause without re-wrapping everything (especially useful in review comments), and it allows easily adding, removing and re-ordering clauses using standard editor functionality. Of course, a set column width is usually somewhat better for lengthy _reading_ sessions, but so are countless other things in the rendered output, which exists for that purpose. See the [SemBr](https://sembr.org/) spec for more.

Reply:
This is what I usually do:
* I don't go on a newline unless the rendered output also has to go (one reason is that by reading the source it might look like a new paragraph, but in the rendered output it isn't).
* Even if I have an 80-char per line limit, I tend to leave some extra margin and stay under 70, both to make it easier to read and to give some room to refactor without having to reflow the whole paragraph.  This is also recommended here:  https://peps.python.org/pep-0008/#maximum-line-length
* I try to break lines in a way that improves readability.  For example, if I can only fit the first couple of words of a new sentence at the end of the previous one, I might just go on a newline, but if I can fit several I don't.
* I don't worry too much about having neat diffs -- if the changes I made don't fit in the extra margin mentioned above I just reflow lines as needed.  If the paragraphs are shorts (and they generally should), you won't usually need to reflow more than a few lines anyway.
* I use a double space after full stops to separate sentences in a paragraph.

The suggestion above is compatible with these guidelines, however if you prefer this style and use it consistently in the sqlite3 docs, I'm fine with it.

Reply:
I'm struggling to see what advantage that approach has over SemBr, other than simple familiarity and marginally better UX when reading a lot of text in source form (which is far from ideal regardless). When I was first introduced to the idea, I was rather skeptical, and particularly concerned about the potential for extra cognitive load when writing them, but using them in them in practice they've ended up feeling very natural after a short amount of time and only require a small amount of additional effort, while bringing significant benefits as a writer and particularly as an editor, wrangler and reviewer.

> I don't go on a newline unless the rendered output also has to go 

Huh? I'm not sure what you mean to say here.

> I don't worry too much about having neat diffs -- if the changes I made don't fit in the extra margin mentioned above I just reflow lines as needed. If the paragraphs are shorts (and they generally should), you won't usually need to reflow more than a few lines anyway.

Well, its not just about `git diff` output being more useful—it benefits anyone who is looking at GitHub or CLI diff, blame, commit/show, PR, etc output by showing only the actual meaningful changes without any distracting reflowing noise. 

Further, it is particularly beneficial for PR reviewers, one of the areas (especially on `cpython`) that is most critically limited for committer bandwidth. It makes it easier to see exactly what significant changes a PR made (especially since reflows break GitHub's otherwise very-useful rich diffs), and makes GH suggestions much more useful, as they can be made just on the logical units affected (typically a single physical line, rather than having to select multiple) with greatly decreased need to worry about reflows, and suggestions on adjacent logical units do not conflict, and can instead be cleanly separated.

Not to mention, it makes it much faster for me to scan the syntactic and semantic structure of the text and consider it in logical units, and avoid missing things when I'm reviewing the source syntax (and I of course read the final rendered result, as the reader would, when reading for overall content). As someone who spends a large amount of my volunteer time reviewing docs and similar content and makes heavy use of suggestions as a critical tool for doing so, this has made a particularly significant improvement to my overall reviewer productivity and efficiency and made reviewing less of a chore.

Reply:
> This is potentially misleading, so I try to avoid adding extra newlines in the source that are not reflected in the rendered output.

Most markup languages work that way. HTML, Markdown, reST, etc. I'd assume that kind of knowledge from folks who work on the docs. Maybe I should not assume that.

> I use a double space after full stops to separate sentences in a paragraph.

Recently (the last year or so) I've seen core devs arguing to stop using that practise; I'll see if I can dig up a link to the discussion. Anyway, using semantic breaks makes that point moot :)

Reply:
> I like the source to visually match the output whenever possible (WYSIWYG).

Yes, but reST (and Markdown, etc) is of course _not_ a WYSIWYG language, and single line breaks in the source are completely ignored in and will almost never match the rendered output anyway in the real world (not to mention all the other processing that reST does that may cause rendered line length to vary). Your much more realistic example above

```rst
We can verify that the new table has been created by querying the
``sqlite_master`` table built-in to SQLite, which should now contain
an entry for the ``movie`` table definition (see `The Schema Table`_
for details).  Execute that query by calling :meth:`cur.execute(...)
<Cursor.execute>`, assign the result to ``res``, and call
:meth:`res.fetchone() <Cursor.fetchone>` to fetch the resulting row::
```

is rendered as (at full width on desktop with our current theme, which may of course change substantially at different widths/devices, switching themes, changing font scaling, etc):

![image](https://user-images.githubusercontent.com/17051931/184864753-8d2fba74-7365-44e9-b990-0a4116870e94.png)

What's more, this is not a bug but a _feature_ :wink: , as it gives the raw source the flexibility to follow a more logical line break convention better suited to editing, reviewing, suggesting and use with source control (the primary purpose of such), while automatically reflowing the rendered output for an optimum experience for readers. Arbitrary fixed hard wrapping gives up most of those benefits, without much of a clear gain that I can see.

> For example, if in the source I write:
> 
> ```
> First line.
> Second line.
> Third line.
> ```

Those are some uncomfortably short consecutive lines for real-world text, which SemBr merely draws attention to. A more realistic example, like that above, will not generally have the psuedo-"WYSIWYG" property illustrated here.

> This is potentially misleading, so I try to avoid adding extra newlines in the source that are not reflected in the rendered output.

I'm not sure I see a categorical distinction in this respect between newlines inserted for hard breaks and those for semantic breaks, as neither are reflected in the output, except that the latter are motivated by the logical content of the text while the former is by arbitrary convention. If readers aren't familiar with such a basic property of most markup languages, there is likely a whole lot else more important they aren't familiar with either.

In fact, maybe its just me, but the fact that SemBr source doesn't (falsely) pretend to look exactly like rendered output actually helps _avoid_ confusion over these breaks being significant, and avoid my typographer's eye naturally picking up on and wanting to fix problems like hanging chads and awkward splices that of course have no relationship to the actual rendered text (to note, professional designers will often actually do their best to massage text to be closer to semantic breaks if practical within the bounds of their target column width, as it reads better).

> If I see something like this in the source I will wonder if the author meant to have two separate paragraphs but forgot the additional newline between the two or not:
> 
> ```
> This is a very long sentence that barely fits on a single
> line.
> This is another sentence that in the source looks like a
> separate paragraph, but  in the output is not.
> ```

But this isn't at all SemBr; in general you'd rarely if ever have a SemBr physical line anywhere near that short, because in general a logical line shouldn't usually be that short, at least with any regularity, or it likely just points to an underlying issue with the structure of the writing itself. Furthermore, given SemBr breaks at logic units, over a passage of non-trivial length the true motivation for the breaks quickly becomes clear, just like for hard breaks. Here's how that would look in actual SemBr (with the standard 79 character max width):

```text
This is a very long sentence that barely fits on a single line.
This is another sentence that in the source looks like a separate paragraph,
but in the output is not.
```

> The link doesn't appear to work for me (both now and when you posted it).

I confirmed it worked when I posted it, but it seems the server is suddenly down now. [Here's the Wayback Machine version](https://web.archive.org/web/20220809231647/https://sembr.org/).

Reply:
I'll mark this as resolved. If needed, we can continue the discussion regarding semantic breaks on Discourse :)